experimental/list_formatter/src/list_formatter.rs

```diff
@@ -0,0 +1,223 @@
+}
+
+impl<'a> ListFormatter<'a> {
+    pub fn new(locale: &str, type_: Type, width: Width) -> Result<ListFormatter<'static>, Error> {
+        match get_patterns(locale, type_, width) {
+            None => Err(Error::UnknownLocale),
+            Some(raw) => { 
+                // let [first, pair, middle, last] = raw.map(Pattern::parse);
+                // Ok(ListFormatter{first, pair, middle, last})
+                Ok(ListFormatter {
+                    first: Pattern::parse(raw[0]), 
+                    pair: Pattern::parse(raw[1]), 
+                    middle: Pattern::parse(raw[2]), 
+                    last: Pattern::parse(raw[3])
+                })
+            }
+        }
+    }
+
+    fn format_internal<B>(
+        &self,
+        values: &[&str],
+        empty: fn() -> B,
+        single: fn(&str) -> B,
+        apply_pattern: fn(&str, &PatternParts<'a>, B) -> B,
+    ) -> B {
+        match values.len() {
+            0 => empty(),
+            1 => single(values[0]),
+            2 => apply_pattern(values[0], &self.pair.get_parts(values[1]), single(values[1])),
+            n => {
+                let mut builder = apply_pattern(values[n-2], &self.last.get_parts(values[n-1]), single(values[n-1]));
+                for i in (1..n - 2).rev() {
+                    builder = apply_pattern(values[i], &self.middle.get_parts(values[i+1]), builder);
+                }
+                apply_pattern(values[0], &self.first.get_parts(values[1]), builder)
+            }
+        }
+    }
+
+    pub fn format(&self, values: &[&str]) -> String {
+        self.format_internal(
+            values,
+            || "".to_string(),
+            |value| value.to_string(),
+            |value, (before, between, after), mut builder| { 
+                builder = builder + after;
+                builder.insert_str(0, between);
+                builder.insert_str(0, value);
+                builder.insert_str(0, before);
+                builder
+            }
+        )
+    }
+
+    pub fn format_to_parts(&self, values: &[&str]) -> SimpleFormattedStringBuilder<FieldType> {
+        self.format_internal(
+            values,
+            SimpleFormattedStringBuilder::<FieldType>::new,
+            |value| {
+                let mut builder = SimpleFormattedStringBuilder::<FieldType>::new();
+                builder.append(value, FieldType::Element);
+                builder
+            },
+            |value, (before, between, after), mut builder| {
+                builder.append(after, FieldType::Literal);
+                builder.prepend(between, FieldType::Literal);
+                builder.prepend(value, FieldType::Element);
+                builder.prepend(before, FieldType::Literal);
+                builder
+            }
+        )
+    }
+}
+
+type PatternParts<'a> = (&'a str,&'a str,&'a str);
+enum Pattern<'a> {
+    Simple(PatternParts<'a>),
+    Conditional { cond: Regex, then: Box<Pattern<'a>>, else_: Box<Pattern<'a>> },
```

Comment:
I do want to follow a data-driven approach though instead of coding up the rules here. Ideally the data provider would expose the logic as a fn(&str) -> bool and the regex parsing would happen somehow before runtime. Do you know if this is possible?